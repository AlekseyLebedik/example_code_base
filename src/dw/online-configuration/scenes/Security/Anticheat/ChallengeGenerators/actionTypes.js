export const CHALLENGE_GENERATORS_FETCH = 'security/CHALLENGE_GENERATORS_FETCH';
export const CHALLENGE_GENERATORS_FETCH_SUCCESS =
  'security/CHALLENGE_GENERATORS_FETCH_SUCCESS';

export const CHALLENGE_GENERATOR_ADD = 'security/CHALLENGE_GENERATOR_ADD';

export const CHALLENGE_GENERATOR_UPDATE = 'security/CHALLENGE_GENERATOR_UPDATE';

export const CHALLENGE_GENERATOR_DELETE = 'security/CHALLENGE_GENERATOR_DELETE';
