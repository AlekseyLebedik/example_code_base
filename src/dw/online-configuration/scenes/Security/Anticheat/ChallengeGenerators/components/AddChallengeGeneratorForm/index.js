import AddChallengeGeneratorForm from './container';

export default AddChallengeGeneratorForm;
