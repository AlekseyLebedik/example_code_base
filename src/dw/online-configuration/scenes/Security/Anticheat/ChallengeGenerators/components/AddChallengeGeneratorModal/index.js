import AddChallengeGeneratorModal from './container';

export default AddChallengeGeneratorModal;
