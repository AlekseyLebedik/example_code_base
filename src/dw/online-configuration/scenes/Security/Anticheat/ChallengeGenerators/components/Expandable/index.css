.common-expandable__container {
  position: relative;
  overflow: hidden;
  display: flex;
  flex-direction: column;
}

.common-expandable__container.more_horiz {
  max-height: 80px;
}

.common-expandable__container.expand_less {
  height: 100%;
  max-height: 300px;
}

.common-expandable__container .common-expandable__wrapper {
  flex-grow: 1;
  overflow: hidden;
}

.common-expandable__container button.common-expandable__button {
  min-height: 24px;
  align-self: center;
}
