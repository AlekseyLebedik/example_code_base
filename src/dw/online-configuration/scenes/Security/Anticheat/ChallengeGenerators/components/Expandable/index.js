import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Icon from '@material-ui/core/Icon';
import IconButton from '@material-ui/core/IconButton';

import Tooltip from '@material-ui/core/Tooltip';

import './index.css';

const initialState = props => ({
  expanded: props.expanded,
});

class Expandable extends Component {
  static propTypes = {
    children: PropTypes.node,
    // eslint-disable-next-line
    expanded: PropTypes.bool,
  };

  static defaultProps = {
    children: undefined,
    expanded: false,
  };

  state = initialState(this.props);

  onToggle = () => {
    this.setState(prevState => ({ expanded: !prevState.expanded }));
  };

  render() {
    const icon = this.state.expanded ? 'expand_less' : 'more_horiz';
    const tooltip = this.state.expanded ? 'Show Less' : 'Show More';
    return (
      <div className={`common-expandable__container ${icon}`}>
        <span className="common-expandable__wrapper">
          {this.props.children}
        </span>
        <Tooltip title={tooltip}>
          <IconButton
            onClick={this.onToggle}
            className="common-expandable__button"
          >
            <Icon>{icon}</Icon>
          </IconButton>
        </Tooltip>
      </div>
    );
  }
}

export default Expandable;
