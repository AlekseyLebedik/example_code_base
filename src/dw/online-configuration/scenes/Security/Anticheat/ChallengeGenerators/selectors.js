import { createSelector } from 'reselect';
import { formatDateTimeSelector } from 'dw/core/helpers/date-time';

const generatorsSelector = state =>
  state.Scenes.Security.Anticheat.ChallengeGenerators.challengeGenerators;

export const challengeGeneratorsSelector = createSelector(
  generatorsSelector,
  formatDateTimeSelector,
  (challengeGenerators, formatDateTime) =>
    challengeGenerators.map(challengeGenerator => ({
      ...challengeGenerator,
      config: JSON.stringify(challengeGenerator.config, null, ' '),
      history: JSON.stringify(challengeGenerator.history, null, ' '),
      lastRun: challengeGenerator.lastRun
        ? formatDateTime(challengeGenerator.lastRun)
        : '',
      nextRun: challengeGenerator.nextRun
        ? formatDateTime(challengeGenerator.nextRun)
        : '',
    }))
);
