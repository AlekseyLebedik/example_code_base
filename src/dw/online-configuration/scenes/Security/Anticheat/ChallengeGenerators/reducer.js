import * as AT from './actionTypes';

const INITIAL_STATE = {
  challengeGenerators: [],
};

const reducer = (state = INITIAL_STATE, action) => {
  switch (action.type) {
    case AT.CHALLENGE_GENERATORS_FETCH_SUCCESS:
      return {
        ...state,
        challengeGenerators: action.challengeGenerators,
      };
    default:
      return state;
  }
};

export default reducer;
