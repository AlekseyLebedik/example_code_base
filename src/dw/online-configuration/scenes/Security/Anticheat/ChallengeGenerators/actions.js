import { nonCriticalHTTPError } from '@demonware/devzone-core/helpers/errors';
import { CriticalErrorActions } from 'dw/core/components/CriticalError';

import { actions as GlobalProgressActions } from '@demonware/devzone-core/components/GlobalProgress';
import * as GlobalSnackBarActions from '@demonware/devzone-core/components/GlobalSnackBar/actions';
import ModalHandlers from 'dw/core/components/ModalHandlers';
import * as AT from './actionTypes';

export const fetchChallengeGenerators = () => ({
  type: AT.CHALLENGE_GENERATORS_FETCH,
});

export const fetchChallengeGeneratorsSuccess = payload => ({
  type: AT.CHALLENGE_GENERATORS_FETCH_SUCCESS,
  challengeGenerators: payload.data,
});

export const fetchChallengeGeneratorsFailed = (err, action) => dispatch => {
  dispatch(GlobalProgressActions.done());
  dispatch(CriticalErrorActions.show(err, () => action));
};

export const addChallengeGenerator = (values, resolve, reject) => dispatch => {
  dispatch({
    type: AT.CHALLENGE_GENERATOR_ADD,
    values,
    resolve,
    reject,
  });
  dispatch(ModalHandlers.submit());
};

export const addChallengeGeneratorSuccess = () => dispatch => {
  dispatch(ModalHandlers.close());
  dispatch(
    GlobalSnackBarActions.show(
      'ChallengeGenerator was successfully added.',
      'success'
    )
  );
  dispatch(fetchChallengeGenerators());
};

export const addChallengeGeneratorFailed = err => dispatch => {
  if (err !== undefined) dispatch(nonCriticalHTTPError(err));
  dispatch(ModalHandlers.stopSubmitting());
};

export const updateChallengeGenerator = values => ({
  type: AT.CHALLENGE_GENERATOR_UPDATE,
  values,
});

export const updateChallengeGeneratorSuccess = values => dispatch => {
  const action = values.state === 'DISABLED' ? 'disabled' : 'enabled';
  dispatch(
    GlobalSnackBarActions.show(
      `Challenge Generator was successfully ${action}.`,
      'success'
    )
  );
  dispatch(fetchChallengeGenerators());
};

export const updateChallengeGeneratorFailed = err => dispatch => {
  dispatch(nonCriticalHTTPError(err));
};

export const deleteChallengeGenerator = generatorId => ({
  type: AT.CHALLENGE_GENERATOR_DELETE,
  generatorId,
});

export const deleteChallengeGeneratorSuccess = () => dispatch => {
  dispatch(
    GlobalSnackBarActions.show(
      'Challenge Generators deleted properly.',
      'success'
    )
  );
  dispatch(fetchChallengeGenerators());
};

export const deleteChallengeGeneratorFailed = err => dispatch => {
  dispatch(nonCriticalHTTPError(err));
};
