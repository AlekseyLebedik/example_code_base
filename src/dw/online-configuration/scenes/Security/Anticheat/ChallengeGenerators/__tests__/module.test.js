import { CriticalErrorActions } from 'dw/core/components/CriticalError';
import { nonCriticalHTTPError } from '@demonware/devzone-core/helpers/errors';
import * as GlobalSnackBarActions from '@demonware/devzone-core/components/GlobalSnackBar/actions';
import ModalHandlers from 'dw/core/components/ModalHandlers';
import * as actions from '../actions';
import * as AT from '../actionTypes';
import reducer from '../reducer';
import { challengeGeneratorsSelector } from '../selectors';

jest.mock('@demonware/devzone-core/components/GlobalSnackBar/actions');
jest.mock('dw/core/components/ModalHandlers');
jest.mock('dw/core/components/CriticalError');
jest.mock('@demonware/devzone-core/helpers/errors');

const generatorId = 1;
const genearator = {
  lastRun: 1531727704,
  generatorId,
  nextRun: 1531742400,
  state: 'DISABLED',
  config: {
    type: 'checksum_blob',
    challenge_group: '1',
    cron: '0 */6 * * *',
    function_id: '200',
    data: {
      min_offset: '0',
      max_offset: '1024',
    },
  },
  history: [],
};
const data = [genearator];
const values = {
  config:
    '{"cron": "0 */6 * * *", "function_id": 200, "challenge_group": 1, "user_group": 0, "data": {"min_offset": 0, "max_offset": 1024}}',
  generatorId: '2',
};
const err = 'error';
const editValues = {
  generatorId,
  state: 'IDLE',
};

describe('Security - Anticheat - ChallengeGenerators', () => {
  describe('Action Creators', () => {
    const MOCKED_CRITICAL_ERROR = 'MOCKED_CRITICAL_ERROR';
    const MOCKED_NON_CRITICAL_ERROR = 'MOCKED_NON_CRITICAL_ERROR';
    const MOCKED_GLOBAL_SNACK_BAR = 'MOCKED_GLOBAL_SNACK_BAR_SHOW';
    const MOCKED_MODAL_HANDLERS_CLOSE = 'MOCKED_MODAL_HANDLERS_CLOSE';
    const MOCKED_MODAL_HANDLERS_SUBMIT = 'MOCKED_MODAL_HANDLERS_SUBMIT';
    const MOCKED_MODAL_HANDLERS_STOP_SUBMITTING =
      'MOCKED_MODAL_HANDLERS_STOP_SUBMITTING';

    beforeAll(() => {
      CriticalErrorActions.show.mockReturnValue({
        type: MOCKED_CRITICAL_ERROR,
      });
      GlobalSnackBarActions.show.mockReturnValue({
        type: MOCKED_GLOBAL_SNACK_BAR,
      });
      ModalHandlers.close.mockReturnValue({
        type: MOCKED_MODAL_HANDLERS_CLOSE,
      });
      ModalHandlers.submit.mockReturnValue({
        type: MOCKED_MODAL_HANDLERS_SUBMIT,
      });
      ModalHandlers.stopSubmitting.mockReturnValue({
        type: MOCKED_MODAL_HANDLERS_STOP_SUBMITTING,
      });
      nonCriticalHTTPError.mockReturnValue({
        type: MOCKED_NON_CRITICAL_ERROR,
      });
    });

    it('CHALLENGE_GENERATORS_FETCH', () => {
      expect(actions.fetchChallengeGenerators()).toHaveProperty(
        'type',
        AT.CHALLENGE_GENERATORS_FETCH
      );
    });

    it('CHALLENGE_GENERATORS_FETCH_SUCCESS', () => {
      const action = actions.fetchChallengeGeneratorsSuccess({ data });

      expect(action).toHaveProperty(
        'type',
        AT.CHALLENGE_GENERATORS_FETCH_SUCCESS
      );
      expect(action).toHaveProperty('challengeGenerators', data);
    });

    it('fetchChallengeGeneratorsFailed', () => {
      const action = actions.fetchChallengeGeneratorsFailed();

      expect(action).toAsyncDispatch({ type: MOCKED_CRITICAL_ERROR });
    });

    it('CHALLENGE_GENERATOR_ADD', () => {
      const resolve = jest.fn();
      const reject = jest.fn();
      const action = actions.addChallengeGenerator(values, resolve, reject);

      expect(action).toAsyncDispatch(
        { type: AT.CHALLENGE_GENERATOR_ADD, values, resolve, reject },
        { type: MOCKED_MODAL_HANDLERS_SUBMIT }
      );
    });

    it('addChallengeGeneratorSuccess', () => {
      const action = actions.addChallengeGeneratorSuccess();

      expect(action).toAsyncDispatch(
        { type: MOCKED_MODAL_HANDLERS_CLOSE },
        { type: MOCKED_GLOBAL_SNACK_BAR },
        { type: AT.CHALLENGE_GENERATORS_FETCH }
      );
    });

    it('addChallengeGeneratorFailed', () => {
      const action = actions.addChallengeGeneratorFailed(err);

      expect(action).toAsyncDispatch(
        { type: MOCKED_NON_CRITICAL_ERROR },
        { type: MOCKED_MODAL_HANDLERS_STOP_SUBMITTING }
      );
    });

    it('CHALLENGE_GENERATOR_UPDATE', () => {
      const action = actions.updateChallengeGenerator(editValues);

      expect(action).toHaveProperty('type', AT.CHALLENGE_GENERATOR_UPDATE);
      expect(action).toHaveProperty('values', editValues);
    });

    it('updateChallengeGeneratorSuccess', () => {
      const action = actions.updateChallengeGeneratorSuccess(editValues);

      expect(action).toAsyncDispatch(
        { type: MOCKED_GLOBAL_SNACK_BAR },
        { type: AT.CHALLENGE_GENERATORS_FETCH }
      );
    });

    it('updateChallengeGeneratorFailed', () => {
      const action = actions.updateChallengeGeneratorFailed(err);

      expect(action).toAsyncDispatch({ type: MOCKED_NON_CRITICAL_ERROR });
    });

    it('CHALLENGE_GENERATOR_DELETE', () => {
      const action = actions.deleteChallengeGenerator(generatorId);

      expect(action).toHaveProperty('type', AT.CHALLENGE_GENERATOR_DELETE);
      expect(action).toHaveProperty('generatorId', generatorId);
    });

    it('deleteChallengeGeneratorSuccess', () => {
      const action = actions.deleteChallengeGeneratorSuccess();

      expect(action).toAsyncDispatch(
        { type: MOCKED_GLOBAL_SNACK_BAR },
        { type: AT.CHALLENGE_GENERATORS_FETCH }
      );
    });

    it('deleteChallengeGeneratorFailed', () => {
      const action = actions.deleteChallengeGeneratorFailed(err);

      expect(action).toAsyncDispatch({ type: MOCKED_NON_CRITICAL_ERROR });
    });
  });

  describe('Reducer', () => {
    it('return the initial state', () => {
      expect(reducer(undefined, {})).toMatchSnapshot();
    });

    it('handle CHALLENGE_GENERATORS_FETCH_SUCCESS', () => {
      expect(
        reducer(undefined, actions.fetchChallengeGeneratorsSuccess({ data }))
      ).toMatchSnapshot();
    });
  });

  describe('Selectors', () => {
    it('validate challengeGeneratorsSelector', () => {
      const state = {
        user: { profile: { timezone: 'America/Vancouver' } },
        Scenes: {
          Security: {
            Anticheat: {
              ChallengeGenerators: { challengeGenerators: data },
            },
          },
        },
      };
      const expected = [
        {
          lastRun: 'Jul 16, 2018 12:55 am PDT',
          generatorId: 1,
          nextRun: 'Jul 16, 2018 05:00 am PDT',
          state: 'DISABLED',
          config:
            '{\n "type": "checksum_blob",\n "challenge_group": "1",\n "cron": "0 */6 * * *",\n "function_id": "200",\n "data": {\n  "min_offset": "0",\n  "max_offset": "1024"\n }\n}',
          history: '[]',
        },
      ];

      expect(challengeGeneratorsSelector(state)).toEqual(expected);
    });
  });
});
