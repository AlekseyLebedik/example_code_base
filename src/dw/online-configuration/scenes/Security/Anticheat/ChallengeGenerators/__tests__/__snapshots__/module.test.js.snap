// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Security - Anticheat - ChallengeGenerators Reducer handle CHALLENGE_GENERATORS_FETCH_SUCCESS 1`] = `
Object {
  "challengeGenerators": Array [
    Object {
      "config": Object {
        "challenge_group": "1",
        "cron": "0 */6 * * *",
        "data": Object {
          "max_offset": "1024",
          "min_offset": "0",
        },
        "function_id": "200",
        "type": "checksum_blob",
      },
      "generatorId": 1,
      "history": Array [],
      "lastRun": 1531727704,
      "nextRun": 1531742400,
      "state": "DISABLED",
    },
  ],
}
`;

exports[`Security - Anticheat - ChallengeGenerators Reducer return the initial state 1`] = `
Object {
  "challengeGenerators": Array [],
}
`;
